File: middle_node.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ptr_slow = head
        ptr_fast = head
        while ptr_fast and ptr_fast.next:
            ptr_slow = ptr_slow.next
            ptr_fast = ptr_fast.next.next
        return ptr_slow

File: js/middleNode.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    ptr1 = head;
    ptr2 = head;
    flipper = false;
    while(ptr2) {
        ptr2 = ptr2.next;
        if (flipper) {
            flipper = false;
            ptr1 = ptr1.next;
        }
        else {
            flipper = true;
        }
        if(!ptr2) {
            return ptr1
        }
    }
    return ptr1
    
};